Fix up locale.c when no mbtowc() nor mbrtowc()

khwilliamson opened the pull request:
This code was written when we still required only a C89 compiler for non-Windows systems without nl_langinfo().  Now that we require C99, mbtowc() and mbrtowc() should be available, and this code could be jettisoned.  But I'd like to get it in good shape before I forget the nuances, so that if it needs to be brought back (say because of a non-conforming compiler or buggy implementation), that it easily can.

I therefore re-examined the code and this series of commits is the result.  Notably the lines affected by its inclusion or not are reduced to just a single area of `locale.c`, to make it easier to revert the removal.  And I figured out a few ways to improve it and save some cycles.

khwilliamson committed:
locale.c: Convert to conditional ? : operator

I think this makes it less clumsy.

khwilliamson committed:
locale.c: Avoid some mallocs

By reusing this buffer, we don't have to realloc unless the next thing
to store in it is bigger than the first.  The order of calling already
has abbreviations (hence shorter) coming after their full names.

khwilliamson committed:
locale.c: Remove potential infinite recursive call

In Configurations where this #ifdef'd code is compiled, we recursively
call my_langinfo().  Prior to this commit, the call asked for the
UTFness of the returned string.  Depending on the particular values
involved, that could lead to this same code being executed to determine
that UTF8ness of the locale.  This would have proceeded ad infinitum
except a previous commit had created flags so as to skip any call that
would recurse infinitely.  But that can lead to erroneous results,
because when skipped, we may not know what the answer is.

This commit avoids all that by not asking the recursed call to return
the UTF8ness of the string, but instead use a heuristic to get its value
here.  This avoids needing to know the locale's UTF8ness (which is where
the infinite recursion would come from).  The heuristic is that if it is
illegal UTF-8, it isn't UTF-8; if it is plain ASCII, we can't tell; and
if it is legal UTF-8, it will be tentatively considered UTF-8.  This is
just one iteration of a loop through a bunch of strings, so that after
all the accumulated evidence of all iterations, we have confidence that
the total result is correct.

There are other code sections that also have the potential for infinite
recursion.  The next commits handle these.

khwilliamson committed:
locale.c: Differently avoid infinite recursion

This commit changes the mechanism for avoiding potential infinite
recursion in my_localeconv().

Normally the UTF8ness of the locale is determined.  Then all strings
returned by localeconv() are examined to see if their SVs need to be
marked as UTF-8 or not.  Knowing that the locale is or isn't UTF-8 helps
in that determination.

But in figuring that out, some Configurations call this function asking
for just a single item.  That would lead to infinite recursion.

To avoid that, on such Configurations prior to this commit, the
UTF8ness of the overall locale wasn't calculated, but instead each
item's UTF8ness was calculated individually.  It's complicated, but it
turns out doing this finesses the issue.  See below for a fuller
explanaation.

This commit changes things so that for single item calls, the UTF8-ness
isn't determined here, but the caller does it itself, and it doesn't
generally need the locale's UTF8ness to make that determination.

To expand on why it's complicated:  This situation arises only on
Configurations where calculating the UTF8ness of the locale may not be
reliable.  But it very likely is reliable except for English locales
whose currency symbol is plain ASCII, such as the USA and Canada and
other former members of the British Empire who use the dollar sign for
their currency symbol.  (I told you it was complicated.)  But for such
locales, the strings are going to all be ASCII, so they aren't going to
be UTF-8, so we don't need to know the locale's UTF8ness.  What both the
previous mechanism and this new one share is both use the function
get_locale_string_utf8ness_i(), and that function has the intelligence
to not need the locale's UTF8ness for an ASCII string.

khwilliamson committed:
locale.c: Move code to earlier and use its result

Under most Configurations, we can determine a locale's code set (like
"ISO 88591-1") directly.  In a few, we have to use more complicated
means, one of which is to see if the name of the locale conforms to the
XPG standard, which includes the code set as part of the name.

It turns out that in most instances, all we care about is if the code
set is UTF-8 or not, and there are ways to do that even on locale names
that don't meet the XPG standard.  However, those ways aren't foolproof.
Especially with English locales whose currency symbol is ASCII (like a
dollar sign or a string of ASCII characters), the code can't make a
definitive determination, and chooses the incorrect answer.  In those
cases, if the locale name does meet XPG criteria, we could use that
as a tie breaker to get the correct answer.

That's what this commit does.  It moves the extraction of the code set
from the locale name to prior to the UTF8ness determination, and when
there is ambiguity (as in the English locales), it uses the found code
set  name to resolve the ambiguity.

khwilliamson committed:
locale.c: White space, comments only

Recent commits have removed the containing blocks for these two sections
of code, so outdent them

khwilliamson committed:
locale.c: Move a couple lines to earlier in the file

This attaches the comment more closely to what it is commenting.

khwilliamson committed:
locale.c: Remove some unnecessary code

This code was a relic which had been lately only necessary to prevent
potential infinite recursion.  The recursion has been removed, so this
can too.

khwilliamson committed:
locale.c: Use MB_CUR_MAX to resolve ambiguities

This code is only compiled for rare Configurations where two
C99-required functions aren't available.  This likely means that someone
ran Configure using arguments to deny their use, which in turn means they
have been found to be buggy on this platform.

The code was originally developed before we required C99, but now
functions as a workaround for their absence, and the code doesn't give
perfect results.  In particular, it fails for English UTF-8 locales
whose names don't meet the XPG standard and which use the dollar sign as
the currency symbol (or other string of ASCII characters).

But, using another C99 feature, MB_CUR_MAX, fixes that.  Our experience
in the field is that it works well.

khwilliamson committed:
locale.c: Use fewer CPU cycles

Now that we have the code set name (if any) before having calculated the
UTF-8ness of the locale (from the past few commits), we can skip some
calculations.  It is very unlikely to be coincidental for the name to be
"UTF-8" and a string from that locale to be syntactically legal
UTF-8 (unless the string is all ASCII).  This is because of the highly
restrictive syntax of UTF-8.  Thus, if we find this situation, we can
presume that the name is telling the truth, and we don't have to keep
checking all the possible strings that we previously did.

tonycoz commented on locale.c:
```diff
@@ -6300,28 +6300,37 @@ S_my_langinfo_i(pTHX_
         utf8ness_t is_utf8 = UTF8NESS_UNKNOWN;
         char * scratch_buf = NULL;
         Size_t scratch_buf_size = 0;
+        const U8 * first_variant;
+        Size_t len;
```
I think these belong within the loops that use them.

There's no harm in having separate declarations of `len` within each loop, and it makes it easier to reason about the use of `len`, since it would not available outside those inner blocks.

Similarly for `first_variant`.

I don't see any uses of `len` or `first_variant` outside of their respective loops, or in the rest of the commit series, but I may have missed them.

I believe this would also allow `len` to be const.

khwilliamson replied:
It's because not both are loops, so the 2nd  does conflict with the first.

But I'm not happy with the duplicated logic; it was ok before these commits, as it was just a line or two.  But it has expanded.  I've now pushed a revised version which has one loop that iterates over all of the tested strings.  It makes a portion more complicated, but the actual logic is squashed into one instance

tonycoz commented on locale.c:
```diff
@@ -6283,21 +6283,100 @@ S_my_langinfo_i(pTHX_
+         * language[_territory[.codeset]][@modifier]
+         *
+         * So we take the part between the dot and any '@' */
+        retval = (const char *) strchr(locale, '.');
```
Why cast here?

In C strchr() returns `char *` which trivially converts to `const char *`.

In C++, since `locale` is `const char *` so the `const char *` [overload from](https://en.cppreference.com/w/cpp/string/byte/strchr)\*:
```
const char* strchr( const char* str, int ch ); 
char* strchr( char* str, int ch );
```
is selected here, which already returns `const char *`.

There's no cast for the similar assignment to `modifier` below.

\* `#include <string.h>` in C++ behaves like including `<cstring.h>` but adds the names that would normally be added to the `std::` namespace to the [global namespace](https://en.cppreference.com/w/cpp/header#C_compatibility_headers)

khwilliamson replied:
It was old code that I simply moved from elsewhere in the file without changing it.  I imagine there were different types at one time that required the case.  But now changed.

khwilliamson committed:
locale.c: Collapse near duplicate code; removes recursion

We do the same operation on both an LC_MONETARY string and various LC_TIME
strings.  Previously the latter was changed to not have the potential
for infinite recursion.  The goal is to make both instances the same
here.  The way this commit does that is to have both instances share the
same code path.  Previously the operation in each was simple, but now it
is more complicated, with further revisions to come.

This however entails setting up some data structures to cope with the
difference in locale categories, which this commit does.

khwilliamson committed:
locale.c: Rename variable

This was shadowing an outer variable, and conflating two things.  We are
looking for the UTF8ness of some strings in a locale to try to divine if
the locale itself is a UTF-8 one or not.  But we're doing this in the
context of trying to find the CODESET of the locale, like 8859-1 or
UTF-8..  And the utf8ness of the CODESET name is always going to be an
ASCII string.  Thus there are two types of utf8ness being looked at
here, and the names of the variables for each should be distinct.

tonycoz commented on locale.c:
```diff
@@ -6294,129 +6255,307 @@ S_my_langinfo_i(pTHX_
+        /* Effectively out-of-bounds, as here there are no trial entries at
+         * all.  This allows this code to compile, but there are no strings to
+         * test, and so the answer will always be non-UTF-8. */
+        const locale_category_index  cat_index =
+                                     follow_on_cat_index = LC_ALL_INDEX_;
```
What declares `follow_on_cat_index` in this case?

Also, `cat_index` is const here, the loop below may never actually execute but it still tries to modify cat_index.

I suspect this should have been:
```
        locale_category_index  cat_index = LC_ALL_INDEX_;
        const locale_category_index  follow_on_cat_index = LC_ALL_INDEX_;
```

khwilliamson replied:
Yes, good catch.  I hadn't bothered to actually build this configuration; but I did with your fixes, and it passed its tests

khwilliamson commented:
Applied via a15d3a085745f320ab9b75bdf56b647a9f9914f9
